### PR TITLE
AAQ-128: Add security to admin app and manage content endpoints

### PR DESCRIPTION
**Pull request by sidravi1**
Reviewer: @ziakhan04 
Estimate: 2 hrs

----

# AAQ-128: Add security to admin app and manage content endpoints

image

## Ticket

Fixes: [AAQ-128](https://idinsight.atlassian.net/browse/AAQ-128?atlOrigin=eyJpIjoiYTdmMWYyODY1MmE4NDVhNWFmNTYxYjQ0NDkyMTEzZWMiLCJwIjoiaiJ9)

## Description, Motivation and Context

Secure admin app and content management endpoints

### Changes

**Backend**
* Add oauth bearer token
* Add login and token generation
* Add two users - "readonly" and "fullaccess"
  * "readonly" cannot edit, create, or delete content. "fullaccess" can.
* Added tests

**Admin app:**
* Added login screen
* Added logout to nav bar
* Added conditionally showing edit, add, and delete buttons based on access level
* Added token retrieval and storage
* Expired or missing token takes you to login screen

## How Has This Been Tested?
- New tests for backend
- Frontend test cases run manually
  - login as "readonly"
    - confirm can see cards and content 
    - confirm no edit/delete/add options
    - logout -> should redirect to login page
  - login as "fullaccess"
    - confirm can see cards and content
    - confirm can edit/delete/add
    - logout -> should redirect to login page
  - open root "localhost:3000/" in incognito window
    - Notice that redirected to login page

## Checklist:

This checklist is a useful reminder of small things that can easily be forgotten.
Put an `x` in all the items that apply and remove any items that are not relevant to this PR.

- [x] My code follows the style guidelines of this project
- [x] I have reviewed my own code to ensure good quality
- [x] I have tested the functionality of my code to ensure it works as intended
- [x] I have resolved merge conflicts
- [x] I have updated the automated tests (if applicable)
- [x] I have written [good commit messages][1]
- [x] I have updated the README file (if applicable)
- [x] I have updated affected documentation (if applicable)

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: admin_app/src/components/ContentCard.tsx [hunk 1]
```diff
@@ -29,21 +30,25 @@ export const ContentCard: React.FC<ContentCardProps> = (
...
-            })
-          }
-        />
-        <TrashIcon
-          className="w-3 h-3 text-gray-500 cursor-pointer hover:text-red-200"
-          onClick={() => card.deleteMe(card.content_id)}
-        />
+        {card.showEditButton ? (
+          <>
+            <PencilIcon
```
Comment (ziakhan04) [inline admin_app/src/components/ContentCard.tsx:35]: Opinion: The Edit and Delete buttons could be twice as big. Right now, it is a bit difficult to click on them.

File: admin_app/src/app/page.tsx [hunk 2]
```diff
@@ -1,9 +1,9 @@
 "use client";
```
Comment (ziakhan04) [inline admin_app/src/app/page.tsx:2]: Aesthetics: Similar to login page, during light mode, the background becoms white for me but then the edit pop-up is black/blue. Good colour combos in both light and dark would look nice I think.

**Review by ziakhan04 — APPROVED** [review]
Thank you! Approving :D

I have made comments about the following (Some of the aesthetic ones + file naming conflict could be addressed in this PR):
- Figure out user management
- Fix a few weird aesthetic problems
- Naming conflict of auth.py files (we have two of them, good to have different names?)

Also, I made a commit to fix some typing issue that I saw which did not allow my app to compile. This was in the content home page .tsx file. ChatGPT help fix it quickly. Apparently, "e.target" needs to be specified a type otherwise Typescript does not compile. I faced this issue when I did docker compose.

Screenshot 2023-11-09 at 7 40 46 PM

Otherwise, I checked all the routes and the frontend functionality. They all work!

File: core_backend/app/routers/auth.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
+from fastapi import APIRouter, Depends, HTTPException
```
[the remark this answers is not in the document]
Comment (sidravi1) [reply core_backend/app/routers/auth.py:1]: makes sense. I'll change this

File: admin_app/src/app/login/page.tsx [hunk 4]
```diff
@@ -0,0 +1,96 @@
+"use client";
```
[the remark this answers is not in the document]
Comment (sidravi1) [reply admin_app/src/app/login/page.tsx:1]: Yep! We have [AAQ-119](https://idinsight.atlassian.net/browse/AAQ-119) to improve styles. Maybe next sprint

File: admin_app/src/components/ContentCard.tsx (hunk 1 above)
Comment (sidravi1) [reply admin_app/src/components/ContentCard.tsx:35]: Added to [AAQ-119](https://idinsight.atlassian.net/browse/AAQ-119)

File: admin_app/src/app/page.tsx (hunk 2 above)
Comment (sidravi1) [reply admin_app/src/app/page.tsx:2]: Added to [AAQ-119](https://idinsight.atlassian.net/browse/AAQ-119) to improve styles

File: core_backend/app/auth.py [hunk 5]
```diff
@@ -1,17 +1,125 @@
...
+oauth2_scheme = OAuth2PasswordBearer(tokenUrl="login")
+
+
+USERS = {
+    "fullaccess": {
+        "password": CONTENT_FULLACCESS_PASSWORD,
+        "access_level": "fullaccess",
+    },
+    "readonly": {"password": CONTENT_READONLY_PASSWORD, "access_level": "readonly"},
+}
```
[the remark this answers is not in the document]
Comment (sidravi1) [reply core_backend/app/auth.py:32]: Yes! We need to a lot more for user management. Created ticket [AAQ-152](https://idinsight.atlassian.net/browse/AAQ-152) to track this.

**Comment by sidravi1** [comment]
Thanks @ziakhan04.

> Also, I made a commit to fix some typing issue that I saw which did not allow my app to compile. This was in the content home page .tsx file. ChatGPT help fix it quickly. Apparently, "e.target" needs to be specified a type otherwise Typescript does not compile. I faced this issue when I did docker compose.

Weird that this didn't show up in local running and only during docker compose. I'll look into why. Thanks for the fix!

**Pull request merged**